### PR TITLE
Add Vendored Libdbus Support

### PR DESCRIPTION
**Pull request by landhb**
This PR adds vendoring behind a feature flag, much like many other `sys` crates in the ecosystem.

- Allows users to build without having `libdbus-dev-1` installed.  Since it will be compiled from source.
- Allows users to cross compile with `cross` without needing a custom docker image.
- Uses the `cc` crate directly, so neither `cmake`  or `meson` are required for downstream users.

A downstream user would simply enable the `vendored` feature:

```
dbus = {version = "0.9.7", features = ["vendored"]}
```

And build with cross:

```
cross build --target x86_64-unknown-linux-musl
```

The PR is currently passing the existing test suite on x86_64:

```
cargo test --features vendored
```

### CODE REVIEW

**Comment by KillingSpark** [comment]
Without being an expert I am just wondering how the licensing works on vendored code, especially since some parts of libdbus are gpl only licenced?

Vendoring this might make it necessary to either strip all gpl Code out if possible and use the afl license or dual licence this as Apache 2 (if not using the vendoring flag) and gpl if using it.

**Comment by landhb** [comment]
I'm definitely now a lawyer/expert either when it comes to license issues.

Best I could find is this answer: 

https://softwareengineering.stackexchange.com/questions/240633/is-licensing-an-issue-for-git-submodules

Which makes it sound like it's not much of an issue.

File: libdbus-sys/build.rs [hunk 1]
```diff
@@ -1,8 +1,320 @@
 extern crate pkg_config;
+use std::error::Error;
+
+// Constants and gated use statements
+#[cfg(feature = "vendored")]
+mod consts {
```
Comment (diwic) [inline libdbus-sys/build.rs:6]: There are a lot of repeated #[cfg(feature = "vendored")] in this file, how about splitting it to a separate file and include it somehow, e g through `#[path = "buildvendored.rs"] mod buildvendored;`

File: libdbus-sys/build.rs [hunk 2]
```diff
@@ -1,8 +1,320 @@
...
+
+#[cfg(feature = "vendored")]
+use consts::*;
+
+/// Source Files Required
+#[cfg(feature = "vendored")]
+const SOURCE_FILES: &'static [&'static str] = &[
+    "./vendor/dbus/dbus/dbus-address.c",
+    "./vendor/dbus/dbus/dbus-auth.c",
+    "./vendor/dbus/dbus/dbus-bus.c",
```
Comment (diwic) [inline libdbus-sys/build.rs:31]: A lot of code in this file might be copy-pasted and slightly changed from `Makefile.am` and `configure.ac`. I'm a bit worried that it makes the code a bit fragile, e g when libdbus updates and adds another source code file the vendoring will break. 

OTOH, trying to autogenerate things might be overkill...or what do you think?

File: .gitmodules [hunk 3]
```diff
@@ -0,0 +1,3 @@
+[submodule "libdbus-sys/vendor/dbus"]
+	path = libdbus-sys/vendor/dbus
+	url = https://github.com/freedesktop/dbus.git
```
Comment (diwic) [inline .gitmodules:3]: Would it be possible to pin this to a specific tag in the repository to reduce the risk of the vendoring build breaking when the dbus.git repo changes?

File: libdbus-sys/build.rs [hunk 4]
```diff
@@ -1,8 +1,320 @@
...
+        // Always sync
+        #define DBUS_USE_SYNC 1
+
+        // Unecessary
+        #undef HAVE_GETPEERUCRED
+        #undef HAVE_CMSGCRED
+    ")?;
+
+    // Define iteger types.
+    //
```
Comment (diwic) [inline libdbus-sys/build.rs:173]: iteger -> integer

**Review by diwic — COMMENTED** [review]
Hi @landhb and thanks for the PR! This seems like a big improvement in usability for the cross compilation use case and I really like that :-)

Apart from the review comments below, I think we should also add a Github CI/workflow for this so it doesn't break without anyone noticing.

**Comment by landhb** [comment]
@diwic No problem! 100% agreed on the testing, I left the PR in draft form since I also wanted to add this to CI/CD.

I also need to make sure it builds on multiple architectures. So far I've only tested x86_64.

**Comment by diwic** [comment]
@KillingSpark from https://github.com/freedesktop/dbus/blob/master/COPYING

"dbus is licensed to you under your choice of the Academic Free
License version 2.1, or the GNU General Public License version 2
(or, at your option any later version)."

From my layman's understanding, neither the AFL (which is a BSD-style license), nor the GNU GPL, make any significant differences depending on whether you link statically or dynamically.

**Comment by KillingSpark** [comment]
> From my layman's understanding, neither the AFL (which is a BSD-style license), nor the GNU GPL, make any significant differences depending on whether you link statically or dynamically. 

I don't know about BSD but I think for GPL it matters for the thing about being derivative work? If it's linked statically your binary is a derivative work but if it's linked dynamically it could be combined with any abi compatible library.

But just as you I am a layman. Just wanted to raise awareness that this might be a topic to think about :)

**Comment by KillingSpark** [comment]
> https://softwareengineering.stackexchange.com/questions/240633/is-licensing-an-issue-for-git-submodules

This discussion is not completely the same. This pr doesn't just use libdbus as a tool while building. I'd agree that that would be fine. It is pulling in and building the code and is necessary for using the library after it is built, making it, in my non-lawyer opinion, either a combined or derivative work.

**Comment by diwic** [comment]
> I don't know about BSD but I think for GPL it matters for the thing about being derivative work?

Again in layman terms:

If you pick GPL for libdbus, and you pick MIT for dbus-rs, then the entire application will most likely be a GPL derivative. 
If you pick AFL for libdbus, and either APL or MIT for dbus-rs, then the entire application can be proprietary, open source, etc, because these licenses are not contagious.

How we link to the libdbus code does not change this.

(LGPL does make a difference between static and dynamic linking, but that does not matter here because libdbus is not licensed under LGPL, and neither is dbus-rs.)

File: libdbus-sys/build.rs (hunk 1 above)
Comment (landhb) [reply libdbus-sys/build.rs]: Good idea! Moved everything over to `build_vendored.rs` in cf25d34a9625a8b3df82b7d2686a428484a41afa.

File: libdbus-sys/build.rs (hunk 2 above)
Comment (landhb) [reply libdbus-sys/build.rs]: Yeah it was taken directly from their [Meson config](https://github.com/freedesktop/dbus/blob/master/dbus/meson.build#L31).

I think combined with pinning to a specific libdbus release (discussed in the other comment) it shouldn't break user's builds. 

So `libdbus-sys` 0.2.3 will be pinned to `libdbus` 1.15.2. And upgrading to `libdbus` 1.15.3 would require updating and testing a new `libdbus-sys` release. So broken builds should be caught during testing before we roll out another version of `libdbus-sys`.

The benefits of using the `cc` crate directly instead of calling `configure`/`cmake`/`meson` is that it doesn't swap the `libdbus-dev-1` dependency for other build-time dependencies. And allows downstream crates/users to simply use cargo to build.

File: .gitmodules (hunk 3 above)
Comment (landhb) [reply .gitmodules:3]: The submodule is actually pinned to the following git tag/release:

```
$ git submodule status
 ed866a94889e13c83dc873d8b5f86a907f908456 libdbus-sys/vendor/dbus (dbus-1.15.2)
 ```
 
 For some reason git doesn't show that in the `.gitmodules` file, only in the git object tree itself.

File: libdbus-sys/build.rs (hunk 4 above)
Comment (landhb) [reply libdbus-sys/build.rs]: Fixed https://github.com/landhb/dbus-rs/blob/vendored/libdbus-sys/build_vendored.rs#L164 thanks!

File: .gitmodules (hunk 3 above)
Comment (diwic) [reply .gitmodules:3]: I looked it up; specifying a sha or tag is not supported, so for stability I suggest we go with the 1.14 branch. You can specify `branch = 1.14` in .gitmodules. When the 1.16 branch is stable I'll be happy to take a PR for that, but the 1.15 branch is a development branch which is more likely to break from time to time, both w r t build failures and runtime bugs.

Does that make sense?

Comment (landhb) [reply .gitmodules:3]: The git-submodule is already tagged at their 1.15.2 release though. If you clone this branch of `dbus-rs` you'll be pinned at `dbus-1.15.2`.

And when we publish to crates.io it will use that tag. Similar to other sys crates, like https://github.com/alexcrichton/openssl-src-rs.

Current tag: https://github.com/freedesktop/dbus/releases/tag/dbus-1.15.2

So I don't think we need to pin to a development branch.

Comment (landhb) [reply .gitmodules:3]: Oh I see what you mean, 1.15 is a development release. Yeah we can pin to the stable 1.14 instead.

Comment (landhb) [reply .gitmodules:3]: Okay now pinned at 1.14.4. Completed in dbc134d22e78aa47cd9f7bb62c4c2e1387c2150f and 4b252a8ce0831ade5473f77b7271aa83c2400ec2.

File: libdbus-sys/build_vendored.rs [hunk 5]
```diff
@@ -0,0 +1,370 @@
...
+const DBUS_VERSION_MINOR: u32 = 15;
+const DBUS_VERSION_MICRO: u32 = 2;
+
+// Location of dbus-daemon socket
+const DBUS_SYSTEM_BUS_DEFAULT_ADDRESS: &str = "/run/dbus/system_bus_socket";
+const DBUS_SESSION_SOCKET_DIR: &str = "/tmp";
+const DBUS_MACHINE_UUID_FILE: &str = "/etc/machine-id";
+
+/// Source Files Required
+const SOURCE_FILES: &'static [&'static str] = &[
```
Comment (diwic) [inline libdbus-sys/build_vendored.rs:26]: I was thinking, to make it as easy as possible to update this file when necessary,
Can you put references on each section? So `SOURCE_FILES` would be:
```
// From vendor/dbus/dbus/meson.build
const DBUS_LIB_SOURCES: &'static [&'static str] = &[
```

...having the same name of the constant is preferred, if this is not practical then add a comment what sections you combined. It does not have to be a step-by-step explanation or anything, just a quick one-line comment.

**Comment by landhb** [comment]
@diwic The `x86_64-unknown-linux-gnu` cross build should be fixed now. That cross image has a much older gcc which didn't understand some of the `pragma` directives.

File: libdbus-sys/build_vendored.rs (hunk 5 above)
Comment (landhb) [reply libdbus-sys/build_vendored.rs:26]: Yeah that's a good idea. Will do.

Comment (landhb) [reply libdbus-sys/build_vendored.rs:27]: Added in https://github.com/diwic/dbus-rs/pull/408/commits/0f1afd0b962a1d9e6043c97bff98fedd26c7933f

**Comment by diwic** [comment]
@landhb Thanks, this is getting better and better :-)

Just to understand how this is all going to work for the end user, will the dbus source end up on crates.io and downloaded with the package? Or will `cross` somehow download sources from git?

File: libdbus-sys/build_vendored.rs [hunk 6]
```diff
@@ -0,0 +1,389 @@
...
+    "./vendor/dbus/dbus/dbus-file-unix.c",
+    "./vendor/dbus/dbus/dbus-pipe-unix.c",
+    "./vendor/dbus/dbus/dbus-sysdeps-pthread.c",
+    "./vendor/dbus/dbus/dbus-sysdeps-unix.c",
+    "./vendor/dbus/dbus/dbus-transport-unix.c",
+    "./vendor/dbus/dbus/dbus-userdb.c",
+];
+
+/// Taken directly from the libdbus meson config
+const CWARNINGS: &'static [&'static str] = &[
```
Comment (diwic) [inline libdbus-sys/build_vendored.rs:96]: Should also have reference

File: libdbus-sys/build_vendored.rs (hunk 5 above)
Comment (landhb) [reply libdbus-sys/build_vendored.rs:28]: And variable names in https://github.com/diwic/dbus-rs/pull/408/commits/94f8e697916d397b34fc44b77f5a433265c87c0b

File: libdbus-sys/build_vendored.rs [hunk 7]
```diff
@@ -0,0 +1,389 @@
...
+use std::collections::hash_map::Iter;
+
+// Version of vendored libdbus
+const DBUS_VERSION: &str = "\"1.14.4\"";
+const DBUS_VERSION_MAJOR: u32 = 1;
+const DBUS_VERSION_MINOR: u32 = 14;
+const DBUS_VERSION_MICRO: u32 = 4;
+
+// Location of dbus-daemon socket
+const DBUS_SYSTEM_BUS_DEFAULT_ADDRESS: &str = "/run/dbus/system_bus_socket";
```
Comment (diwic) [inline libdbus-sys/build_vendored.rs:21]: These three constants should also have a reference

File: libdbus-sys/cross_compile.md [hunk 8]
```diff
@@ -13,6 +13,22 @@ The examples below all assume you're trying to compile for Raspberry Pi 2 or 3 r
 Cross compiling using rust-embedded/cross
 =========================================
 
+#### New Instructions:
```
Comment (diwic) [inline libdbus-sys/cross_compile.md:16]: Skip this line

File: libdbus-sys/cross_compile.md [hunk 9]
```diff
@@ -13,6 +13,22 @@ The examples below all assume you're trying to compile for Raspberry Pi 2 or 3 r
...
+
+**Update**: With the new `vendored` feature. It is no longer required to provide a custom docker image, or have `libdbus-1-dev` installed to build. Simply enable the `vendored` feature:
```
Comment (diwic) [inline libdbus-sys/cross_compile.md:18]: Change to something like: 

It is not required to provide a custom docker image, or have `libdbus-1-dev` installed to build. Git will need to be installed on the build host in order to download the dbus sources. Then enable the `vendored` feature:

File: libdbus-sys/build_vendored.rs [hunk 10]
```diff
@@ -0,0 +1,389 @@
...
+}
+
+/// Config.h and dbus-arch-deps.h stubs (we generate config programatically
+/// instead to be passed directly to cc)
+fn generate_stubs() -> Result<(), Box<dyn Error>> {
+    let mut config = File::create("vendor/dbus/config.h")?;
+    let mut arch = File::create("vendor/dbus/dbus/dbus-arch-deps.h")?;
+
+    // Stub out the file.
+    config.write_all(b"
```
Comment (diwic) [inline libdbus-sys/build_vendored.rs:202]: Needs reference (if there is one?)

File: libdbus-sys/build_vendored.rs [hunk 11]
```diff
@@ -0,0 +1,389 @@
...
+
+    // Define integer types.
+    //
+    // The libdbus builder relies on detecting int/long/long long width.
+    // but we can just use modern C types like uint64_t.
+    //
+    // Pointer width will still be detected at build time by our
+    // generate_config() method.
+    arch.write_all(b" 
+        #ifndef DBUS_ARCH_DEPS_H
```
Comment (diwic) [inline libdbus-sys/build_vendored.rs:223]: Needs reference (if there is one?)

File: libdbus-sys/build_vendored.rs [hunk 12]
```diff
@@ -0,0 +1,389 @@
...
+
+        DBUS_END_DECLS
+
+        #endif /* DBUS_ARCH_DEPS_H */
+    ")?;
+    Ok(())
+}
+
+/// Generate configuration based on target platform + architecture
+fn generate_config(cc: &mut cc::Build, config: &mut Config) -> Result<(), Box<dyn Error>> {
```
Comment (diwic) [inline libdbus-sys/build_vendored.rs:255]: Reference would be nice here too

File: libdbus-sys/build_vendored.rs [hunk 13]
```diff
@@ -0,0 +1,389 @@
...
+/// enough to port to cc directly. Allowing downstream crates to rely on the
+/// dbus crate without additional host dependencies.
+pub fn build_libdbus() -> Result<(), Box<dyn Error>> {
+    let mut compiler = cc::Build::new();
+    let mut config = Config::new();
+
+    // Ensure submodule is checked out
+    if !Path::new("vendor/dbus").exists() {
+        let _ = Command::new("git")
+            .args(&["submodule", "update", "--init", "vendor/dbus"])
```
Comment (diwic) [inline libdbus-sys/build_vendored.rs:346]: Hmm. Does this not mean that git the binary is now a build dependency of `dbus-rs`, if the vendored feature is enabled?

Comment (diwic) [reply libdbus-sys/build_vendored.rs:346]: If installing git is a step required for cross build then it should be documented.

**Review by diwic — CHANGES_REQUESTED** [review]
Okay, I hope this is the final set of review comments. After fixing, please squash the commits into one and I believe it should be ready for merging.

File: libdbus-sys/build_vendored.rs (hunk 13 above)
Comment (landhb) [reply libdbus-sys/build_vendored.rs:342]: @diwic When published on `crates.io` it won't be required.  This is more of a development convenience for people who have cloned the repository without `--recursive`.

Using `libgit2-sys` as an example.

They have the C source as a git submodule here: https://github.com/rust-lang/git2-rs/tree/master/libgit2-sys
Then when published on crates.io it's archived with the full source at the specific commit: https://docs.rs/crate/libgit2-sys/0.14.0+1.5.0/source/libgit2/

The build.rs line is something they also have, but isn't explicitly required. When cargo pulls down the `libdbus-sys` source it will include the full `dbus` source as well. So the `git` line won't run since `vendor/dbus` exists.

Their build.rs: https://github.com/rust-lang/git2-rs/blob/master/libgit2-sys/build.rs#L27

**Comment by landhb** [comment]
> @landhb Thanks, this is getting better and better :-)
> 
> Just to understand how this is all going to work for the end user, will the dbus source end up on crates.io and downloaded with the package? Or will `cross` somehow download sources from git?
> 
> Edit: Found the answer myself - the build system calls `git`.

@diwic So my understanding is that publishing on `crates.io` archives the entire git submodule along with the source. Elaborated more [here](https://github.com/diwic/dbus-rs/pull/408#discussion_r1057796424) with `libgit2-sys` as an example.

File: libdbus-sys/cross_compile.md (hunk 8 above)
Comment (landhb) [reply libdbus-sys/cross_compile.md]: Removed in 728539ee9c1a88ccf61a2c2336b168b2f99f429f

File: libdbus-sys/build_vendored.rs (hunk 7 above)
Comment (landhb) [reply libdbus-sys/build_vendored.rs:22]: Documented in 84b3ddf35a391205ce7340325c5b01879ced1359

File: libdbus-sys/build_vendored.rs (hunk 10 above)
Comment (landhb) [reply libdbus-sys/build_vendored.rs:205]: Currently no reference. The actual config is created dynamically in `generate_config()` but `config.h` is required to exist since some `dbus` source files include it.

File: libdbus-sys/build_vendored.rs (hunk 11 above)
Comment (landhb) [reply libdbus-sys/build_vendored.rs:222]: Also no static reference, when building with `cmake/meson` this is created. And really all it does is setup the types for `dbus_int16_t`, `dbus_int32_t`, and `dbus_int64_t`. Plus their unsigned counterparts.

File: libdbus-sys/build_vendored.rs (hunk 6 above)
Comment (landhb) [reply libdbus-sys/build_vendored.rs:99]: Documented in 84b3ddf35a391205ce7340325c5b01879ced1359

**Comment by diwic** [comment]
@landhb 
So I merged it, then just ran (locally here, from the workspace directory) 

```
$ git pull --rebase origin
$ cargo test --features vendored
error: failed to run custom build command for `libdbus-sys v0.2.2 (/home/david/rust/dbus-rs/libdbus-sys)`

Caused by:
  process didn't exit successfully: `/home/david/rust/dbus-rs/target/debug/build/libdbus-sys-55ef2401f6ef29e5/build-script-build` (exit status: 1)
  --- stdout
  cargo:rerun-if-changed=build.rs
  cargo:rerun-if-changed=build_vendored.rs

  --- stderr
  Error: Os { code: 2, kind: NotFound, message: "No such file or directory" }
```

Inside `libdbus-sys/vendor` there is a single `/dbus/config.h` which is 0 bytes. Any ideas? Or is this only supposed to work with cross, not normal testing/compilation?

**Comment by landhb** [comment]
@diwic Yeah looks like I typo'd the source path check. Submitted a fix here: https://github.com/diwic/dbus-rs/pull/409

**Comment by landhb** [comment]
Thanks for all the review work @diwic :tada:

**Comment by diwic** [comment]
@landhb Thanks for the contribution! Nice to see some improvement in this area :-)

I'll wait a few days to see if any bugs come up as a result of this PR, otherwise I'll prepare a release of libdbus-sys and dbus-rs. Sounds like a plan?

**Comment by diwic** [comment]
@landhb I notice tons of the following warnings in the CI output, anything to worry about?

```

libdbus-sys-linux (stable, aarch64-unknown-linux-musl)
The `set-output` command is deprecated and will be disabled soon. Please upgrade to using Environment Files. For more information see: https://github.blog/changelog/2022-10-11-github-actions-deprecating-save-state-and-set-output-commands/
```

**Comment by landhb** [comment]
Looks like it's the toolchain action: https://github.com/actions-rs/toolchain/issues/221

Since we aren't explicitly using set-output. 

Looks like [dtonly ](https://github.com/dtolnay) has a drop in replacement now, that we can swap for:

https://github.com/marketplace/actions/rustup-toolchain-install

**Comment by landhb** [comment]
@diwic Started a pull here to upgrade some of those deps: https://github.com/diwic/dbus-rs/pull/411

**Comment by diwic** [comment]
@landhb Released now, enjoy!
